compiler/src/iree/compiler/Dialect/Util/Analysis/Constant/ConstExpr.cpp

```diff
@@ -12,36 +12,37 @@
 #include "llvm/Support/CommandLine.h"
 #include "llvm/Support/Debug.h"
 #include "llvm/Support/GraphWriter.h"
-#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Interfaces/SideEffectInterfaces.h"
 
 #define DEBUG_TYPE "iree-constexpr"
 
-using llvm::dbgs;
-
-using namespace mlir::iree_compiler::IREE::Util;
-
 namespace mlir::iree_compiler::IREE::Util {
 
 //===----------------------------------------------------------------------===//
 // ConstExprAnalysis
 //===----------------------------------------------------------------------===//
 
-namespace {
-OpOperand *findOperandFor(Operation *op, Value input) {
+static OpOperand *findOperandFor(Operation *op, Value input) {
   for (OpOperand &operand : op->getOpOperands()) {
     if (operand.get() == input)
       return &operand;
   }
   return nullptr;
 }
 
-bool isHoistableToRootOp(Operation *rootOp, Operation *constOp) {
-  Operation *syms = SymbolTable::getNearestSymbolTable(constOp);
-  return syms == rootOp;
+bool ConstExprAnalysis::isConstExprOperation(Operation *queryOp) const {
+  if (queryOp->getNumResults() == 0) {
+    bool hasNoMemoryEffects = false;
+    if (auto effectOp = dyn_cast<MemoryEffectOpInterface>(queryOp)) {
+      hasNoMemoryEffects = effectOp.hasNoEffect();
+    }
+    if (hasNoMemoryEffects && queryOp->hasTrait<OpTrait::ReturnLike>())
+      return true;
+    return false;
+  }
+  return isConstExprValue(queryOp->getResult(0));
```

Comment:
I honestly don't fully understand why, but it always did this :P I suspect it's because the operations have all their result values inserted into the map with the same value. Something I want to fix in the future.